feat: multiple selection of tracks

Sytronik opened the pull request:
This implements multiple selection by cmd+click, shift+click, shift+arrowup/down

### Multiple selection behaviours

The behaviour is close to that of macOS Finder. (not exactly the same)
**pivot** is needed for shift+click.

1. click
   - the clicked one is the only selection
   - it becomes the pivot
3. cmd+click
   - if the clicked one is *not* already selected
       - add the clicked one to the selection
       - it becomes the pivot
   - otherwise
       - remove the clicked on from the selection
       -  if it was the pivot, the recently selected one becomes the pivot
4. shift+click / shift+arrowup/down
   - the pivot is not changed.
   - first, **remove** all tracks *selected after the pivot*
   - then, add all tracks between the pivot and the clicked one (including the clicked one).
      - For shift+arrowup/down, the clicked one means a track before/after the recently selected one, respectively.

### Changes of behaviour of adding/removing  tracks
1. after adding
   - select *all the added tracks*
2. "delete track" context menu
   - if the menu is popped up on one of the selections, remove *all the selections*
   - otherwise, remove that track only. (**TODO**: we have to indicate this behaviour for users)
3. after removing
   - retain not-removed selection.
   - if no retain exists, the nearest track from the previous track is selected

Sytronik committed:
ArrowUp keycmd selects one before the recent ...

selection, not before the oldest selection

Sytronik committed:
implement multiple selection of tracks

Sytronik committed:
change behaviour of multi-selection when adding/removing tracks

1. after adding: select all the added tracks

2. "delete track" context menu: if the menu is popped up on one of the selections, remove all the selections
if not, remove that track only. (TODO: we have to indicate this behaviour for users)

3. after removing: retian not removed selection. if no retain exists, the nearest track from the previous track is selected

Sytronik committed:
add keycommand cmd(ctrl)+a (select all)

Sytronik committed:
misc code style changes

crackyachae commented on src/renderer/hooks/useSelectedTracks.ts:
```diff
@@ -1,41 +1,113 @@
     e.preventDefault();
 
+    if (isCommand(e)) {
+      const selectedIndexOfId = selectedTrackIds.indexOf(id);
```
현재 id의 selectedTracks 내에서의 index니까 indexInSelectedOfId가 더 낫지 않으려나. 너무 번잡한가?

Sytronik replied:
좀만 줄여서 idxInSelectedIds 이렇게 갈까 그럼

crackyachae commented on src/renderer/hooks/useSelectedTracks.ts:
```diff
@@ -1,41 +1,113 @@
+      if (indexOfId > indexOfPivot) addingIds = trackIds.slice(indexOfPivot + 1, indexOfId + 1);
+      else addingIds = trackIds.slice(indexOfId, indexOfPivot);
+      // if newSelected has some of addingIds, remove them first
+      newSelected = newSelected.filter((selectedId) => !addingIds.includes(selectedId));
```
아래서 추가로 `push` 하는 것 대신 `newSelected = newSelected.filter((selectedId) => !addingIds.includes(selectedId)).concat(addingIds)`로 체이닝 하는 게 더 깔끔할 것 같음.

Sytronik replied:
ok

crackyachae commented on src/renderer/hooks/useSelectedTracks.ts:
```diff
@@ -1,41 +1,113 @@
+    ) {
+      return;
     }
+    setPivotId(trackIds[-1]);
```
`trackIds[trackIds.length - 1]` 로 수정

Sytronik replied:
아 파이썬하던 버릇땜에 ㅋㅋㅋㅋ ㄱㅅ

crackyachae reviewed (commented):
코드 상에서 수정해야 할 부분은 코멘트 남겨두었음.

그 외에 궁금한 점이나 이슈는,
1. 어떤 경우에 macOS와 동일하지 않게 동작하는지
2. cmd를 눌러 여러 트랙을 선택한 뒤 해제할 때 pivot된 트랙이 현재 화면에 보이지 않으면 스크롤이 해당 트랙으로 이동하는데 불편한 것 같음. 자연스러운 동작이면 수정하지 않아도 괜찮을듯.
3. pivot을 4번 트랙이라고 가정했을 때 shift를 누르고 0번 트랙을 눌러서 0 ~ 4를 선택한 뒤, 다시 shift를 누른 채로 2번 트랙(즉, 0번과 4번 사이에 있는 트랙)을 선택하면 0 ~ 1번 트랙은 선택이 해제되어야 하는 데 선택이 해제되지 않음.

Sytronik commented:
1. pivot 0번 기준으로 4번까지 shift로 선택->2번을 cmd로 선택해제-> 이경우 mac은 3번이 pivot이 됨(cmd로 선택해제하면 무조건 가까운 애가 pivot되는 방식인듯)
이 구현에서는 그 동작이 부자연스러운 것 같다고 판단해서 cmd로 (현재 pivot을 선택해제하지 않는 한) 선택해제로는 pivot이 바뀌지 않게 했음. 뭔가 더 복잡한 예외케이스에서도 동작이 다를수 있을것 같아서 일일이 뭐가 다른지 써놓지는 않았음. 일단 일반적으로 많이 쓰는 패턴(shift만 사용/cmd만 사용/shift 후 cmd로 몇개 빼거나 더함)에서는 다 동일하니까 직관성에 위배되는건 없다고 판단함.

2. 이거는 정확히 어떤 케이스에서 스크롤시킬지 논의해서 정책을 정해야 한다고 생각했고 & 스크롤 안하는 예외 케이스를 추가하는건 리팩토링 이후에 해도 괜찮을거라 판단함. issue에 넣어놔야할듯

3. 이건 확인 후 수정해놓겠음

Sytronik committed:
fix wrong indexing

Sytronik committed:
use Array.concat instead of push

Sytronik committed:
fix order of adding selections by shift ...

bug scenario
AS-IS
pivot 3 -> shift+click 0
-> selectedTrackIds: [3,0,1,2]
-> then,  shif+click 2 doesn't work

TO-BE
pivot 3 -> shift_click 0
-> selectedTrackIds: [3,2,1,0]
-> then, shif+cick 2
-> seldctedTrackIds: [3,2]

Sytronik committed:
change var name

Sytronik commented:
Applied all suggestions ...
except fixing bad UX caused by interaction between multi-selection and "auto-scrolling to selected track"

crackyachae reviewed (approved):
lgtm
